finality-aleph/src/network/component.rs

```diff
@@ -10,12 +11,73 @@ pub trait Sender<D: Data>: Sync + Send + Clone {
     fn send(&self, data: D, recipient: Recipient) -> Result<(), SendError>;
 }
 
+#[derive(Clone)]
+pub struct MapSender<D, S> {
+    sender: S,
+    _phantom: PhantomData<D>,
+}
+
+pub trait SenderMap<From: Data>: Sender<From> {
+    fn map(self) -> MapSender<From, Self> {
+        MapSender {
+            sender: self,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<D: Data, S: Sender<D>> SenderMap<D> for S {}
+
+impl<D: Data, S: Sender<D>, IntoD: Data + Into<D>> Sender<IntoD> for MapSender<D, S> {
+    fn send(&self, data: IntoD, recipient: Recipient) -> Result<(), SendError> {
+        self.sender.send(data.into(), recipient)
+    }
+}
+
 /// For receiving arbitrary messages.
 #[async_trait::async_trait]
 pub trait Receiver<D: Data>: Sync + Send {
     async fn next(&mut self) -> Option<D>;
 }
 
+pub struct MapReceiver<D, R> {
+    receiver: R,
+    _phantom: PhantomData<D>,
+}
+
+pub trait ReceiverMap<From: Data>: Receiver<From> + Sized {
+    fn map(self) -> MapReceiver<From, Self> {
+        MapReceiver {
+            receiver: self,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<D: Data, R: Receiver<D>> ReceiverMap<D> for R {}
+
+#[async_trait::async_trait]
+impl<D: Data, R: Receiver<D>, FromD: Data + TryFrom<D>> Receiver<FromD> for MapReceiver<D, R>
+where
+    FromD::Error: Display,
+{
+    async fn next(&mut self) -> Option<FromD> {
+        loop {
+            let data = self.receiver.next().await;
+            let data = match data {
+                Some(data) => data,
+                None => return None,
+            };
+            match TryFrom::try_from(data) {
+                Ok(message) => return Some(message),
+                Err(e) => {
+                    warn!(target: "aleph-network", "Error decoding message in MapReceiver: {}", e)
```

Comment:
should this be an error?

Reply:
It is based on https://github.com/Cardinal-Cryptography/aleph-node/blob/1abe0ac671f4be8c5f0176f1b2dfcd0b0c73fe03/finality-aleph/src/network/service.rs#L221 . I am not sure. Some way of reasoning: error should for some serious problems, like some component can't operate - a user seeing it should be scared :). warn is more like - meh, failed, but lets move on.

Reply:
sounds fair